cpp/pagerank_module/algorithm_online/pagerank.cpp

```diff
@@ -0,0 +1,222 @@
+#include <memory>
+#include <unordered_map>
+
+#include <mg_graph.hpp>
+#include "pagerank.hpp"
+
+namespace pagerank_online_alg {
+namespace {
+PageRankData context;
+std::uint64_t global_R;
+double global_epsilon;
+
+void NormalizeRank(std::vector<std::pair<std::uint64_t, double>> &rank) {
+  const double sum = std::accumulate(rank.begin(), rank.end(), 0.0, [](auto &sum, auto &p) { return sum + p.second; });
+  for (auto &[node_id, value] : rank) {
+    value /= sum;
+  }
+}
+
+std::vector<std::pair<std::uint64_t, double>> CalculatePageRank() {
+  std::vector<std::pair<std::uint64_t, double>> pageranks;
+
+  auto R = pagerank_online_alg::global_R;
+  auto eps = pagerank_online_alg::global_epsilon;
+
+  auto n = pagerank_online_alg::context.walks_counter.size();
+  for (auto const [node_id, total] : pagerank_online_alg::context.walks_counter) {
+    auto rank = total / ((n * R) / eps);
+    pageranks.emplace_back(node_id, rank);
+  }
+
+  NormalizeRank(pageranks);
+  return pageranks;
+}
+
+void CreateRoute(const mg_graph::GraphView<> &graph, const std::uint64_t start_id, std::vector<std::uint64_t> &walk,
+                 const std::uint64_t walk_index, const double epsilon, std::uniform_real_distribution<float> distr,
+                 std::mt19937 gen) {
+  std::uint64_t current_id = start_id;
+  while (true) {
+    auto neighbors = graph.Neighbours(current_id);
+    if (neighbors.empty()) break;
+
+    // Pick and add the random outer edge
+    auto number_of_neighbors = neighbors.size();
+    auto next_id = neighbors[std::rand() % number_of_neighbors].node_id;
+    next_id = graph.GetMemgraphNodeId(next_id);
+
+    walk.emplace_back(next_id);
+    pagerank_online_alg::context.walks_table[next_id].insert(walk_index);
+    pagerank_online_alg::context.walks_counter[next_id]++;
+
+    // Finish walk when random number is smaller than epsilon
+    // Average length of walk is 1/epsilon
+    if (distr(gen) < epsilon) {
+      break;
+    }
+
+    current_id = graph.GetInnerNodeId(next_id);
+  }
+}
+
+void UpdateCreate(const mg_graph::GraphView<> &graph, const std::pair<std::uint64_t, std::uint64_t> new_edge) {
+  auto [from, to] = new_edge;
+
+  std::unordered_set<std::uint64_t> walk_table_copy(pagerank_online_alg::context.walks_table[from]);
+  for (auto walk_index : walk_table_copy) {
+    auto &walk = pagerank_online_alg::context.walks[walk_index];
+
+    auto position = std::find(walk.begin(), walk.end(), from) + 1;
```

Comment:
Is it guaranteed that from is contained inside the walk? 
What if you get vector::end as the result of the find? Adding 1 would give you something that is not equal to vector::end.

Reply:
It is guaranteed, but a priori we do not have the information about a position within the walk. That's why I move to the next position after calling a find in a walk.